loleaflet/src/layer/tile/CanvasTileLayer.js

```diff
@@ -5472,12 +5472,144 @@ L.CanvasTileLayer = L.Layer.extend({
+		for (var i = 0; i < this._parts; i++) {
+			var partMin = i * partHeightPixels;
+			var partMax = (i + 1) * partHeightPixels;
+			if ((yMin >= partMin && yMin <= partMax) || (yMax >= partMin && yMax <= partMax)) {
+				parts.push(i);
+			}
+		}
+
+		for (i = 0; i < parts.length; i++) {
+			var yMinLocal = yMin - parts[i] * partHeightPixels;
+			if (yMinLocal >= 0) {
+				yMinLocal = yMinLocal - yMinLocal % this._tileSize;
+				queue.push(new L.TileCoordData(currentX, yMinLocal, zoom, parts[i]));
+			}
+
+			var yMaxLocal = yMax - parts[i] * partHeightPixels;
+			if (yMaxLocal >= 0) {
+				yMaxLocal = yMaxLocal - yMaxLocal % this._tileSize;
+				if (yMaxLocal !== yMinLocal)
+					queue.push(new L.TileCoordData(currentX, yMaxLocal, zoom, parts[i]));
+			}
+		}
+	},
+
+	_sortFileBasedQueue: function (queue) {
+		for (var i = 0; i < queue.length - 1; i++) {
+			for (var j = i + 1; j < queue.length; j++) {
+				var a = queue[i];
+				var b = queue[j];
+				var switchTiles = false;
+
+				if (a.part === b.part) {
+					if (a.y > b.y) {
+						switchTiles = true;
+					}
+					else if (a.y === b.y) {
+						switchTiles = a.x > b.x;
+					}
+					else {
+						switchTiles = false;
+					}
+				}
+				else {
+					switchTiles = a.part > b.part;
+				}
+
+				if (switchTiles) {
+					var temp = a;
+					queue[i] = b;
+					queue[j] = temp;
+				}
+			}
+		}
+	},
+
+	_updateFileBasedView: function (checkOnly) {
+		if (this._partHeightTwips === 0) // This is true before status message is handled.
+			return [];
+
+		var zoom = Math.round(this._map.getZoom());
+
+		var ratio = this._tileSize / this._tileHeightTwips;
+		var partHeightPixels = Math.round((this._partHeightTwips + this._spaceBetweenParts) * ratio);
+
+		var topLeft = app.sectionContainer.getDocumentTopLeft();
+		var bottomRight = app.sectionContainer.getDocumentBottomRight();
+
+		topLeft = [topLeft[0] - topLeft[0] % this._tileSize, topLeft[1] - topLeft[1] % this._tileSize];
+		bottomRight = [bottomRight[0] - bottomRight[0] % this._tileSize, bottomRight[1] - bottomRight[1] % this._tileSize];
+
+		var currentX = topLeft[0];
+		var currentY = topLeft[1];
+
+		var queue = [];
+		while (currentY <= bottomRight[1]) {
+			if (currentX > bottomRight[0]) {
+				currentX = topLeft[0];
+				currentY += this._tileSize;
+			}
+
+			this._getTileInfoForFileBasedView(currentX, currentY, zoom, partHeightPixels, queue);
+			currentX += this._tileSize;
+		}
+
+		this._sortFileBasedQueue(queue);
+
+		if (checkOnly) {
+			return queue;
+		}
+		else {
+			this._sendClientVisibleArea();
+			this._sendClientZoom();
+
+			for (var i = 0; i < queue.length; i++) {
+				var key = this._tileCoordsToKey(queue[i]);
+				var tile = this._tiles[key];
+				if (!tile) {
+					tile = this.createTile(this._wrapCoords(queue[i]), L.bind(this._tileReady, this, queue[i]));
+
+					// save tile in cache
+					this._tiles[key] = {
+						el: tile,
+						wid: 0,
+						coords: queue[i],
+						current: true
+					};
+
```

Comment:
This section related to tile object vs tile img object, will need to adapt to slurping related changes whereby the tileCache has the resolved image elements and that we don't have to deal with img's src attribute. See commits:
https://github.com/CollaboraOnline/online/commit/663016d5d0a0e7e193d9328d40493fb4e9f18519
https://github.com/CollaboraOnline/online/commit/5e64575ab687e203c38e388a2d165d772e352561

This is also true for co-6-4.